File: backend/app/equipment.py
from fastapi import APIRouter, Depends, HTTPException
from sqlalchemy.orm import Session
from sqlalchemy import text
from typing import List, Optional

from .db import get_db
from .schemas import EquipmentCreate, EquipmentOut, SerialNumberCreate, SerialNumberOut
from .auth import get_current_user

router = APIRouter(prefix="/equipment", tags=["equipment"])


# ==================== EQUIPMENT ====================

@router.get("", response_model=List[EquipmentOut])
def list_equipment(
    skip: int = 0,
    limit: int = 100,
    search: Optional[str] = None,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Получить список оборудования"""
    query = "SELECT * FROM equipment"
    params = {}
    
    if search:
        query += " WHERE material_number ILIKE :search OR name ILIKE :search"
        params["search"] = f"%{search}%"
    
    query += " ORDER BY created_at DESC LIMIT :limit OFFSET :skip"
    params["limit"] = limit
    params["skip"] = skip
    
    result = db.execute(text(query), params)
    return [dict(row._mapping) for row in result]


@router.get("/{equipment_id}", response_model=EquipmentOut)
def get_equipment(
    equipment_id: int,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Получить оборудование по ID"""
    result = db.execute(
        text("SELECT * FROM equipment WHERE id = :id"),
        {"id": equipment_id}
    ).first()
    
    if not result:
        raise HTTPException(status_code=404, detail="Equipment not found")
    
    return dict(result._mapping)


@router.get("/by-material/{material_number}", response_model=EquipmentOut)
def get_by_material_number(
    material_number: str,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Получить оборудование по номеру материала"""
    result = db.execute(
        text("SELECT * FROM equipment WHERE material_number = :mn"),
        {"mn": material_number}
    ).first()
    
    if not result:
        raise HTTPException(status_code=404, detail="Equipment not found")
    
    return dict(result._mapping)


@router.post("", response_model=EquipmentOut)
def create_equipment(
    data: EquipmentCreate,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Создать новое оборудование"""
    # Проверяем уникальность номера материала
    existing = db.execute(
        text("SELECT id FROM equipment WHERE material_number = :mn"),
        {"mn": data.material_number}
    ).first()
    
    if existing:
        raise HTTPException(status_code=400, detail="Material number already exists")
    
    result = db.execute(
        text("""
            INSERT INTO equipment (material_number, name, description, category, unit)
            VALUES (:mn, :name, :desc, :cat, :unit)
            RETURNING *
        """),
        {
            "mn": data.material_number,
            "name": data.name,
            "desc": data.description,
            "cat": data.category,
            "unit": data.unit
        }
    ).first()
    
    db.commit()
    return dict(result._mapping)


# ==================== SERIAL NUMBERS ====================

@router.get("/{equipment_id}/serials", response_model=List[SerialNumberOut])
def list_serial_numbers(
    equipment_id: int,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Получить серийные номера для оборудования"""
    result = db.execute(
        text("""
            SELECT sn.*, e.material_number, e.name as equip_name, e.description as equip_desc,
                   e.category as equip_cat, e.unit as equip_unit, e.created_at as equip_created
            FROM serial_numbers sn
            JOIN equipment e ON sn.equipment_id = e.id
            WHERE sn.equipment_id = :eid
            ORDER BY sn.created_at DESC
        """),
        {"eid": equipment_id}
    )
    
    serials = []
    for row in result:
        d = dict(row._mapping)
        serials.append({
            "id": d["id"],
            "equipment_id": d["equipment_id"],
            "serial_number": d["serial_number"],
            "warehouse_id": d["warehouse_id"],
            "status": d["status"],
            "notes": d["notes"],
            "created_at": d["created_at"],
            "equipment": {
                "id": d["equipment_id"],
                "material_number": d["material_number"],
                "name": d["equip_name"],
                "description": d["equip_desc"],
                "category": d["equip_cat"],
                "unit": d["equip_unit"],
                "created_at": d["equip_created"]
            }
        })
    
    return serials


@router.post("/{equipment_id}/serials", response_model=SerialNumberOut)
def add_serial_number(
    equipment_id: int,
    data: SerialNumberCreate,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Добавить серийный номер к оборудованию"""
    # Проверяем существование оборудования
    equip = db.execute(
        text("SELECT id FROM equipment WHERE id = :id"),
        {"id": equipment_id}
    ).first()
    
    if not equip:
        raise HTTPException(status_code=404, detail="Equipment not found")
    
    # Проверяем уникальность серийного номера
    existing = db.execute(
        text("SELECT id FROM serial_numbers WHERE serial_number = :sn"),
        {"sn": data.serial_number}
    ).first()
    
    if existing:
        raise HTTPException(status_code=400, detail="Serial number already exists")
    
    result = db.execute(
        text("""
            INSERT INTO serial_numbers (equipment_id, serial_number, warehouse_id, status, notes)
            VALUES (:eid, :sn, :wh, :status, :notes)
            RETURNING *
        """),
        {
            "eid": equipment_id,
            "sn": data.serial_number,
            "wh": data.warehouse_id,
            "status": data.status or "available",
            "notes": data.notes
        }
    ).first()
    
    db.commit()
    return dict(result._mapping)


@router.get("/serials/search", response_model=SerialNumberOut)
def search_by_serial(
    serial: str,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Найти оборудование по серийному номеру"""
    result = db.execute(
        text("""
            SELECT sn.*, e.material_number, e.name as equip_name, e.description as equip_desc,
                   e.category as equip_cat, e.unit as equip_unit, e.created_at as equip_created
            FROM serial_numbers sn
            JOIN equipment e ON sn.equipment_id = e.id
            WHERE sn.serial_number = :serial
        """),
        {"serial": serial}
    ).first()
    
    if not result:
        raise HTTPException(status_code=404, detail="Serial number not found")
    
    d = dict(result._mapping)
    return {
        "id": d["id"],
        "equipment_id": d["equipment_id"],
        "serial_number": d["serial_number"],
        "warehouse_id": d["warehouse_id"],
        "status": d["status"],
        "notes": d["notes"],
        "created_at": d["created_at"],
        "equipment": {
            "id": d["equipment_id"],
            "material_number": d["material_number"],
            "name": d["equip_name"],
            "description": d["equip_desc"],
            "category": d["equip_cat"],
            "unit": d["equip_unit"],
            "created_at": d["equip_created"]
        }
    }


@router.patch("/serials/{serial_id}/status")
def update_serial_status(
    serial_id: int,
    status: str,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Обновить статус серийного номера"""
    valid_statuses = ["available", "in_use", "defective", "written_off"]
    if status not in valid_statuses:
        raise HTTPException(status_code=400, detail=f"Invalid status. Valid: {valid_statuses}")
    
    result = db.execute(
        text("UPDATE serial_numbers SET status = :status WHERE id = :id RETURNING *"),
        {"status": status, "id": serial_id}
    ).first()
    
    if not result:
        raise HTTPException(status_code=404, detail="Serial number not found")
    
    db.commit()
    return {"status": "updated", "serial": dict(result._mapping)}


# ==================== EQUIPMENT WITH COUNTS ====================

@router.get("/with-counts/list")
def list_equipment_with_counts(
    search: Optional[str] = None,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Получить список оборудования с общим количеством по каждому типу"""
    query = """
        SELECT 
            e.id, e.material_number, e.name, e.description, e.category, e.unit, e.created_at,
            COUNT(sn.id) as total_serial_count,
            COUNT(sn.id) FILTER (WHERE sn.status = 'available') as available_count,
            COUNT(sn.id) FILTER (WHERE sn.status = 'in_use') as in_use_count,
            COUNT(sn.id) FILTER (WHERE sn.status = 'defective') as defective_count,
            COALESCE(ws.total_stock, 0) as stock_quantity
        FROM equipment e
        LEFT JOIN serial_numbers sn ON sn.equipment_id = e.id AND sn.status != 'written_off'
        LEFT JOIN (
            SELECT equipment_id, SUM(quantity) as total_stock 
            FROM warehouse_stock 
            GROUP BY equipment_id
        ) ws ON ws.equipment_id = e.id
    """
    params = {}
    
    if search:
        query += " WHERE e.material_number ILIKE :search OR e.name ILIKE :search"
        params["search"] = f"%{search}%"
    
    query += " GROUP BY e.id, ws.total_stock ORDER BY e.created_at DESC"
    
    result = db.execute(text(query), params)
    
    items = []
    for row in result:
        d = dict(row._mapping)
        items.append({
            "id": d["id"],
            "material_number": d["material_number"],
            "name": d["name"],
            "description": d["description"],
            "category": d["category"],
            "unit": d["unit"],
            "created_at": d["created_at"],
            "total_count": (d["total_serial_count"] or 0) + (d["stock_quantity"] or 0),
            "serial_count": d["total_serial_count"] or 0,
            "available_count": d["available_count"] or 0,
            "in_use_count": d["in_use_count"] or 0,
            "defective_count": d["defective_count"] or 0,
            "stock_quantity": d["stock_quantity"] or 0
        })
    
    return items


@router.get("/{equipment_id}/warehouse-distribution")
def get_equipment_warehouse_distribution(
    equipment_id: int,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Получить распределение оборудования по складам"""
    # Серийные номера по складам
    serials_by_warehouse = db.execute(text("""
        SELECT w.id as warehouse_id, w.name as warehouse_name, w.is_central,
               COUNT(sn.id) as serial_count,
               COUNT(sn.id) FILTER (WHERE sn.status = 'available') as available_count,
               COUNT(sn.id) FILTER (WHERE sn.status = 'in_use') as in_use_count,
               COUNT(sn.id) FILTER (WHERE sn.status = 'defective') as defective_count
        FROM warehouses w
        LEFT JOIN serial_numbers sn ON sn.warehouse_id = w.id 
            AND sn.equipment_id = :eid AND sn.status != 'written_off'
        GROUP BY w.id, w.name, w.is_central
        HAVING COUNT(sn.id) > 0
        ORDER BY w.is_central DESC, w.name
    """), {"eid": equipment_id})
    
    # Остатки по складам (несерийное)
    stock_by_warehouse = db.execute(text("""
        SELECT w.id as warehouse_id, w.name as warehouse_name, w.is_central,
               ws.quantity
        FROM warehouse_stock ws
        JOIN warehouses w ON w.id = ws.warehouse_id
        WHERE ws.equipment_id = :eid AND ws.quantity > 0
        ORDER BY w.is_central DESC, w.name
    """), {"eid": equipment_id})
    
    distribution = {}
    
    # Добавляем серийные
    for row in serials_by_warehouse:
        d = dict(row._mapping)
        wid = d["warehouse_id"]
        distribution[wid] = {
            "warehouse_id": wid,
            "warehouse_name": d["warehouse_name"],
            "is_central": d["is_central"],
            "serial_count": d["serial_count"],
            "available_count": d["available_count"],
            "in_use_count": d["in_use_count"],
            "defective_count": d["defective_count"],
            "stock_quantity": 0
        }
    
    # Добавляем остатки
    for row in stock_by_warehouse:
        d = dict(row._mapping)
        wid = d["warehouse_id"]
        if wid in distribution:
            distribution[wid]["stock_quantity"] = d["quantity"]
        else:
            distribution[wid] = {
                "warehouse_id": wid,
                "warehouse_name": d["warehouse_name"],
                "is_central": d["is_central"],
                "serial_count": 0,
                "available_count": 0,
                "in_use_count": 0,
                "defective_count": 0,
                "stock_quantity": d["quantity"]
            }
    
    return list(distribution.values())


# ==================== SERIAL SEARCH IMPROVED ====================

@router.get("/serials/search-advanced")
def search_serials_advanced(
    serial: Optional[str] = None,
    equipment_id: Optional[int] = None,
    warehouse_id: Optional[int] = None,
    status: Optional[str] = None,
    limit: int = 50,
    db: Session = Depends(get_db),
    user=Depends(get_current_user)
):
    """Расширенный поиск серийных номеров"""
    query = """
        SELECT sn.*, e.material_number, e.name as equip_name, e.description as equip_desc,
               e.category as equip_cat, e.unit as equip_unit, e.created_at as equip_created,
               w.name as warehouse_name
        FROM serial_numbers sn
        JOIN equipment e ON sn.equipment_id = e.id
        LEFT JOIN warehouses w ON sn.warehouse_id = w.id
        WHERE 1=1
    """
    params = {"limit": limit}
    
    if serial:
        query += " AND sn.serial_number ILIKE :serial"
        params["serial"] = f"%{serial}%"
    
    if equipment_id:
        query += " AND sn.equipment_id = :eid"
        params["eid"] = equipment_id
    
    if warehouse_id:
        query += " AND sn.warehouse_id = :wid"
        params["wid"] = warehouse_id
    
    if status:
        query += " AND sn.status = :status"
        params["status"] = status
    
    query += " ORDER BY sn.created_at DESC LIMIT :limit"
    
    result = db.execute(text(query), params)
    
    serials = []
    for row in result:
        d = dict(row._mapping)
        serials.append({
            "id": d["id"],
            "equipment_id": d["equipment_id"],
            "serial_number": d["serial_number"],
            "warehouse_id": d["warehouse_id"],
            "warehouse_name": d["warehouse_name"],
            "status": d["status"],
            "notes": d["notes"],
            "created_at": d["created_at"],
            "equipment": {
                "id": d["equipment_id"],
                "material_number": d["material_number"],
                "name": d["equip_name"],
                "description": d["equip_desc"],
                "category": d["equip_cat"],
                "unit": d["equip_unit"],
                "created_at": d["equip_created"]
            }
        })
    
    return serials

File: frontend/src/api/equipment.ts
import { api } from './client';
import type { Equipment, EquipmentCreate, SerialNumber, SerialNumberCreate } from '../types';

// ==================== EQUIPMENT ====================

// Получить список оборудования
export const listEquipment = async (search?: string): Promise<Equipment[]> => {
  const params = search ? { search } : {};
  const response = await api.get('/equipment', { params });
  return response.data;
};

// Получить оборудование по ID
export const getEquipment = async (id: number): Promise<Equipment> => {
  const response = await api.get(`/equipment/${id}`);
  return response.data;
};

// Получить оборудование по номеру материала
export const getEquipmentByMaterialNumber = async (materialNumber: string): Promise<Equipment> => {
  const response = await api.get(`/equipment/by-material/${materialNumber}`);
  return response.data;
};

// Создать оборудование
export const createEquipment = async (data: EquipmentCreate): Promise<Equipment> => {
  const response = await api.post('/equipment', data);
  return response.data;
};

// ==================== SERIAL NUMBERS ====================

// Получить серийные номера для оборудования
export const listSerialNumbers = async (equipmentId: number): Promise<SerialNumber[]> => {
  const response = await api.get(`/equipment/${equipmentId}/serials`);
  return response.data;
};

// Добавить серийный номер
export const addSerialNumber = async (equipmentId: number, data: SerialNumberCreate): Promise<SerialNumber> => {
  const response = await api.post(`/equipment/${equipmentId}/serials`, data);
  return response.data;
};

// Найти по серийному номеру
export const searchBySerialNumber = async (serial: string): Promise<SerialNumber> => {
  const response = await api.get('/equipment/serials/search', { params: { serial } });
  return response.data;
};

// Обновить статус серийного номера
export const updateSerialStatus = async (serialId: number, status: string): Promise<SerialNumber> => {
  const response = await api.patch(`/equipment/serials/${serialId}/status`, null, {
    params: { status }
  });
  return response.data.serial;
};

// ==================== ROLES ====================

// Получить список ролей
export const getRoles = async (): Promise<{ value: string; label: string }[]> => {
  const response = await api.get('/roles');
  return response.data.roles;
};

// Получить список пользователей (admin only)
export const listUsers = async (): Promise<{ id: number; username: string; role: string; warehouse_id: number | null; warehouse_name: string | null; is_active: boolean }[]> => {
  const response = await api.get('/users');
  return response.data as { id: number; username: string; role: string; warehouse_id: number | null; warehouse_name: string | null; is_active: boolean }[];
};

// Создать пользователя с ролью (admin only)
export const createUserWithRole = async (username: string, password: string, role: string) => {
  const response = await api.post('/auth/create-user', null, {
    params: { username, password, role }
  });
  return response.data;
};

// Изменить роль пользователя (admin only)
export const updateUserRole = async (userId: number, role: string) => {
  const response = await api.patch(`/users/${userId}/role`, null, {
    params: { role }
  });
  return response.data;
};

// Привязать склад к пользователю (admin only)
export const assignUserWarehouse = async (userId: number, warehouseId: number | null) => {
  const response = await api.patch(`/users/${userId}/warehouse`, null, {
    params: { warehouse_id: warehouseId }
  });
  return response.data;
};

// Изменить статус активности пользователя (admin only)
export const toggleUserActive = async (userId: number, isActive: boolean) => {
  const response = await api.patch(`/users/${userId}/active`, null, {
    params: { is_active: isActive }
  });
  return response.data;
};

// Изменить пароль пользователя (admin only)
export const changeUserPassword = async (userId: number, newPassword: string) => {
  const response = await api.patch(`/users/${userId}/password`, null, {
    params: { new_password: newPassword }
  });
  return response.data;
};

// Удалить пользователя (admin only)
export const deleteUser = async (userId: number) => {
  const response = await api.delete(`/users/${userId}`);
  return response.data;
};

// ==================== INVENTORY ====================

// Переместить оборудование по серийному номеру
export const transferEquipment = async (serialNumber: string, toWarehouseId: number, notes?: string) => {
  const response = await api.post('/inventory/transfer', {
    serial_number: serialNumber,
    to_warehouse_id: toWarehouseId,
    notes
  });
  return response.data;
};

// Добавить материалы на склад
export const addStock = async (equipmentId: number, warehouseId: number, quantity: number, notes?: string) => {
  const response = await api.post('/inventory/add-stock', {
    equipment_id: equipmentId,
    warehouse_id: warehouseId,
    quantity,
    notes
  });
  return response.data;
};

// Списать материалы
export const writeOffStock = async (data: {
  equipment_id: number;
  warehouse_id: number;
  quantity: number;
  serial_number?: string;
  notes?: string;
}) => {
  const response = await api.post('/inventory/write-off', data);
  return response.data;
};

// История перемещений
export const listTransactions = async (limit?: number) => {
  const response = await api.get('/inventory/transactions', {
    params: { limit }
  });
  return response.data;
};

// Остатки на центральном складе
export const getCentralStock = async () => {
  const response = await api.get('/inventory/central-stock');
  return response.data;
};

// Остатки на складе
export const getWarehouseStock = async (warehouseId: number) => {
  const response = await api.get(`/warehouses/${warehouseId}/stock`);
  return response.data;
};

// ==================== EQUIPMENT WITH COUNTS ====================

// Получить список оборудования с количеством
export const listEquipmentWithCounts = async (search?: string) => {
  const params = search ? { search } : {};
  const response = await api.get('/equipment/with-counts/list', { params });
  return response.data;
};

// Получить распределение оборудования по складам
export const getEquipmentWarehouseDistribution = async (equipmentId: number) => {
  const response = await api.get(`/equipment/${equipmentId}/warehouse-distribution`);
  return response.data;
};

// ==================== ADVANCED SERIAL SEARCH ====================

// Расширенный поиск серийных номеров
export const searchSerialsAdvanced = async (params: {
  serial?: string;
  equipment_id?: number;
  warehouse_id?: number;
  status?: string;
  limit?: number;
}) => {
  const response = await api.get('/equipment/serials/search-advanced', { params });
  return response.data;
};

// ==================== BULK TRANSFER ====================

// Массовое перемещение по списку серийных номеров
export const bulkTransferEquipment = async (serialNumbers: string[], toWarehouseId: number, notes?: string) => {
  const response = await api.post('/inventory/bulk-transfer', {
    serial_numbers: serialNumbers,
    to_warehouse_id: toWarehouseId,
    notes
  });
  return response.data;
};
